src/v/storage/tests/compaction_fuzz_test.cc

```diff
@@ -0,0 +1,196 @@
+// Copyright 2024 Redpanda Data, Inc.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.md
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0
+
+#include "base/vlog.h"
+#include "container/fragmented_vector.h"
+#include "model/record_batch_types.h"
+#include "model/timeout_clock.h"
+#include "random/generators.h"
+#include "storage/record_batch_builder.h"
+#include "storage/tests/utils/disk_log_builder.h"
+#include "storage/types.h"
+
+#include <seastar/core/io_priority_class.hh>
+#include <seastar/core/sleep.hh>
+#include <seastar/testing/thread_test_case.hh>
+
+#include <absl/container/btree_map.h>
+#include <boost/test/tools/old/interface.hpp>
+#include <boost/test/unit_test.hpp>
+
+#include <exception>
+
+using namespace std::chrono_literals;
+
+namespace {
+ss::logger cmp_testlog("cmp-fuzz");
+} // anonymous namespace
+
+static model::record_batch make_random_batch(
+  model::offset offset,
+  bool empty,
+  model::record_batch_type type,
+  std::vector<std::optional<ss::sstring>> keys,
+  std::vector<std::optional<ss::sstring>> values,
+  int num_records) {
+    BOOST_REQUIRE(keys.size() == values.size());
+    storage::record_batch_builder builder(type, offset);
+    auto to_iobuf = [](std::optional<ss::sstring> x) {
+        std::optional<iobuf> result;
+        if (x.has_value()) {
+            iobuf buf;
+            buf.append(x->data(), x->size());
+            result = std::move(buf);
+        }
+        return result;
+    };
+    if (!empty) {
+        for (int i = 0; i < num_records; i++) {
+            auto key = random_generators::random_choice(keys);
+            auto val = random_generators::random_choice(values);
+            builder.add_raw_kv(to_iobuf(key), to_iobuf(val));
+        }
+    }
+    return std::move(builder).build();
+}
```

Comment:
is this doing something specific to the test? pretty sure we have at least a few versions of random batch creation utilities in the tree